typedload/exceptions.py

```diff
@@ -0,0 +1,140 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# author Salvo "LtWorf" Tomaselli <tiposchi@tiscali.it>
+
+
+from enum import Enum
+from typing import Any, List, NamedTuple, Optional, Type, Union
+
+
+class AnnotationType(Enum):
+    """
+    The types of annotation, used by different loaders.
+
+    FIELD is the name of a field
+    INDEX is the numerical index of a value, in subscriptable objects-
+    """
+    FIELD = 'field'
+    INDEX = 'index'
+    FORWARDREF = 'forwardref'
+    KEY = 'key'
+    VALUE = 'value'
+    UNION = 'union'
+
+
+Annotation = NamedTuple('Annotation', [
+    ('annotation_type', AnnotationType),
+    ('value', Union[str, int]),
+])
+
+
+TraceItem = NamedTuple('TraceItem', [
+    ('value', Any),
+    ('type_', Type),
+    ('annotation', Optional[Annotation]),
+])
+
+
+class TypedloadException(Exception):
+    """
+    Exception which exposes some extra fields.
+
+    trace:
+        It is a list of all the recursive invocations of load(), with the
+        parameters used.
+        Very useful to locate the issue.
+        The annotation is used by complex loaders that call load() more than
+        once, to indicate in which step the error occurred.
+        For example a list loader will use it to indicate the index which had
+        the exception, and a NamedTuple loader will use it to indicate the name
+        of the field which generated the exception.
+
+    value:
+        contains the value that could not be loaded.
+
+    type_:
+        contains the type in which the value could not be loaded.
+
+    exceptions:
+        A list of exceptions that happened during the loading.
+        This is for now only used by the Union loader, to list all the
+        exceptions that occurred during the various attempts.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+        self.trace = []  # type: List[TraceItem]
+        self.value = kwargs.get('value')  # type: Any
+        self.type_ = kwargs.get('type_')  # type: Optional[Type]
+        self.exceptions = kwargs.get('exceptions', [])  # type: List[Exception]
+
+    def __str__(self) -> str:
+        def compress_value(v: Any) -> str:
+            v = str(v)
+            if len(v) > 80:
+                return v[:77] + '...'
+            return v
+        e = '%s\nValue: %s\nType: %s\n' % (
+            super().__str__(),
+            compress_value(self.value),
+            self.type_
+        )
+        if self.trace:
+            e += '\nLoad trace:\n'
+        path = []  # type: List[str]
+        for i in self.trace:
+            e += 'Type: %s ' % i.type_
+            if i.annotation:
+                e += 'Annotation: (%s %s) ' % (i.annotation[0], i.annotation[1])
+                path.append(i.annotation[1] if type(i.annotation[1]) != int else '[%d]' % i.annotation[1])  # type: ignore
```

Comment:
Why is the `type: ignore` necessary here?

Reply:
```
mypy --config-file mypy.conf typedload
typedload/exceptions.py:104: error: Argument 1 to "append" of "list" has incompatible type "Union[str, int]"; expected "str"
typedload/exceptions.py:104: error: Incompatible types in string interpolation (expression has type "Union[str, int]", placeholder has type "Union[int, float]")
make: *** [Makefile:9: mypy] Error 1

Reply:
It doesn't figure out that as a result of the if, only a string goes there.